pandas/_libs/tslibs/conversion.pyx

```diff
@@ -73,6 +73,123 @@ cdef inline int64_t get_datetime64_nanos(object val) except? -1:
 
     return ival
 
+
+def cast_to_nanoseconds(ndarray arr):
+    """
+    Ensure a np.datetime64 array has dtype specifically 'datetime64[ns]'
+
+    Parameters
+    ----------
+    arr : ndarray
+
+    Returns
+    -------
+    result : ndarray with dtype datetime64[ns]
+
+    """
+    cdef:
+        Py_ssize_t i, n = arr.size
+        ndarray[int64_t] ivalues, iresult
+        PANDAS_DATETIMEUNIT unit
+        pandas_datetimestruct dts
+
+    shape = (<object> arr).shape
+
+    ivalues = arr.view(np.int64).ravel()
+
+    result = np.empty(shape, dtype='M8[ns]')
+    iresult = result.ravel().view(np.int64)
+
+    if len(iresult) == 0:
+        return result
+
+    unit = get_datetime64_unit(arr.flat[0])
```

Comment:
prob should add some checks around this, IOW this could already be ns (TODO ok)

Reply:
Yah, this check is done elsewhere, will not be hard to implement.

If it weren't for `datetime64[Y]` and `datetime64[M]` we could take the `pandas_datetime_to_datetimestruct` out of these funcs altogether, avoid converting back and forth.